gslb/k8sobjects/ingress_object.go

```diff
@@ -82,20 +84,53 @@ func getTLSHosts(ingress *v1beta1.Ingress) []string {
 	return tlsHosts
 }
 
+func parseVSAndControllerAnnotations(annotations map[string]string) (map[string]string, string, error) {
+	vsUUIDs, controllerUUID := make(map[string]string), ""
+	if len(annotations) == 0 {
+		return vsUUIDs, controllerUUID, nil
+	}
+	vsAnnotations, exists := annotations[VSAnnotation]
+	if !exists {
+		gslbutils.Debugf("No VS Annotations exist for object, annotations: %v", annotations)
+		return vsUUIDs, controllerUUID, nil
+	}
+	controllerUUID, exists = annotations[ControllerAnnotation]
+	if !exists {
+		gslbutils.Debugf("No Controller Annotation exist for object, annotations: %v", annotations)
+		return vsUUIDs, controllerUUID, nil
+	}
+	if err := json.Unmarshal([]byte(vsAnnotations), &vsUUIDs); err != nil {
+		return vsUUIDs, controllerUUID, fmt.Errorf("error in unmarshalling VS annotations: %v", err)
+	}
+
+	return vsUUIDs, controllerUUID, nil
+}
+
 // GetIngressHostMeta returns a ingress split into its backends
 func GetIngressHostMeta(ingress *v1beta1.Ingress, cname string) []IngressHostMeta {
 	ingHostMetaList := []IngressHostMeta{}
 	hostIPList := gslbutils.IngressGetIPAddrs(ingress)
 	tlsHosts := getTLSHosts(ingress)
+	vsUUIDs, controllerUUID, err := parseVSAndControllerAnnotations(ingress.Annotations)
+	if err != nil {
+		gslbutils.Errf("error in parsing VS and Controller annotations for ingress %s/%s: %v",
+			ingress.Namespace, ingress.Name, err)
+	}
 	for _, hip := range hostIPList {
+		vsUUID, ok := vsUUIDs[hip.Hostname]
+		if !ok {
+			vsUUID = ""
+		}
 		metaObj := IngressHostMeta{
-			IngName:   ingress.Name,
-			Namespace: ingress.ObjectMeta.Namespace,
-			Hostname:  hip.Hostname,
-			IPAddr:    hip.IPAddr,
-			Cluster:   cname,
-			ObjName:   ingress.Name + "/" + hip.Hostname,
-			TLS:       false,
+			IngName:            ingress.Name,
+			Namespace:          ingress.ObjectMeta.Namespace,
+			Hostname:           hip.Hostname,
+			IPAddr:             hip.IPAddr,
+			Cluster:            cname,
+			ObjName:            ingress.Name + "/" + hip.Hostname,
+			TLS:                false,
+			VirtualServiceUUID: vsUUID,
```

Comment:
what does an empty virtualServiceUUID mean? Is it a valid case or is it an invalid config?
If it's an invalid case, then we could return an `err` from the `parseVSAndControllerAnnotations
` method when we are unable to fetch the vsUUID or the controllerUUID?

Reply:
In the ingestion layer, we don't check for the validity of the ingress, because we store all the objects, either to the accepted or to the rejected cache. However, you are right, this is an error condition and if the user needs VS sync for an ingress object and doesn't have the VS UUID as part of the annotation. This needs to be communicated back to the user. This was previously done in the rest layer, but I have optimized it a bit.

Now, if an ingress has invalid VS/Controller annotations (and the overall cluster requires VS sync), we now handle it layer wise:
1. Ingestion layer: We write in the logs that the ingress doesn't have proper annotations.
2. Graph layer: We won't add/update the member if the VSUUID or Controller UUID is empty (and, syncVipOnly is false for this member) and return with an error.
3. Rest layer: We update the VSUUID and the Controller UUID in the GS object member only if !syncVIPOnly.